pep-0692.rst

```diff
@@ -0,0 +1,344 @@
+
+Current specification enables for type hinting ``**kwargs`` as long as all of the keyword arguments specified by them are of the same type. However, that behaviour can be very limiting. Therefore, in this PEP we propose a new way to enable more precise ``**kwargs`` typing. The new approach revolves around using ``TypedDict`` to type ``**kwargs`` that comprise keyword arguments of different types.
+
+Motivation
+==========
+
+Currently annotating ``**kwargs`` with a type ``T`` means that the ``kwargs`` type is in fact ``Dict[str, T]``. For example:
+
+::
+
+    def foo(**kwargs: str) -> None: ...
+
+means that all keyword arguments in ``foo`` are strings (i.e. ``kwargs`` is of type ``Dict[str, str]``). This behaviour limits the ability to type annotate ``**kwargs`` only to the cases where all of them are of the same type. However, it is often the case that keyword arguments conveyed by ``**kwargs`` have different types that are dependent on the keyword's name. In those cases type annotating ``**kwargs`` is not possible. This is especially a problem for already existing codebases where the need of refactoring the code in order to introduce proper type annotations may be considered not worth the effort. This in turn prevents the project from getting all of the benefits that type hinting can provide. As a consequence, there has been a lot of discussion around supporting more precise ``**kwargs`` typing [#mypyIssue4441]_ and it became a feature that is necessary for a large part of Python community.
+
+Rationale
+=========
+
+PEP 589 introduced the ``TypedDict`` type constructor that supports dictionary types consisting of string keys and values of potentially different types. This makes ``TypedDict`` a perfect candidate to be used for more precise ``**kwargs`` typing. However, specifing ``**kwargs`` type using a ``TypedDict`` means, as mentioned earlier, that each keyword argument specified by ``**kwargs`` is a ``TypedDict`` itself. For instance:
+
+::
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+    
+    def foo(**kwargs: Movie) -> None: ...
+
+means that each keyword argument in ``foo`` is itself a ``Movie`` dictionary that has a ``name`` key with a string type value and a ``year`` key with an integer type value. Therefore, in order to support specifing ``kwargs`` type as a ``TypedDict`` without breaking current behaviour, a new syntax has to be introduced.
+
+Specification
+=============
+
+To support the aforementioned use case we propose to use the double asterisk syntax inside of the type annotation. The required grammar change is discussed in more detail in section `Grammar Changes`_. Continuing the previous example:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+would mean that the ``**kwargs`` comprise two keyword arguments specified by ``Movie`` (i.e. a ``name`` keyword of type ``str`` and a ``year`` keyword of type ``int``). Then, inside the function itself, appropriate type checking would take place:
```

Comment:
"appropriate type checking would take place" at type checking time, not at runtime.